includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,384 @@
+}
+
+/**
+ * Function to send a CURL request
+ * @param string $endpoint The URL which is the target of the request
+ * @param array $headers A list of HTTP headers to send
+ * @param string $content The optional body of the request, if applicable
+ * @param string $type The HTTP method to use (defaults to GET)
+ * @param array $expect An optional keyed array of HTTP response codes mapped to an (optional) expected JSON identifier
+ *                      If provided, this is required for the function to return data
+ * @return mixed A string or array.
+ *               If no $expect parameter is given, returns the response body.
+ *               If an $expect key matches with no array value, returns the response body only on a match
+ *               If an $expect key matches and has a value, returns the decoded JSON as an array if that matched value is present as a key
+ */
+function islandora_video_streaming_panopto_request_curl($endpoint, $headers, $content = '', $type = 'GET', $expect = array()) {
+  $cURL = curl_init();
+  curl_setopt($cURL, CURLOPT_URL, $endpoint);
+  curl_setopt($cURL, CURLOPT_RETURNTRANSFER, true);
+  curl_setopt($cURL, CURLOPT_HEADER, 1);
+  curl_setopt($cURL, CURLINFO_HEADER_OUT, true);
+  switch (strtoupper($type)) {
+    case 'POST':
+    curl_setopt($cURL, CURLOPT_POST, true);
+    break;
+    case 'PUT':
+    curl_setopt($cURL, CURLOPT_CUSTOMREQUEST, 'PUT');
+    break;
+  }
+  if ($content) {
+    curl_setopt($cURL, CURLOPT_POSTFIELDS, $content);
+  }
+  curl_setopt($cURL, CURLOPT_HTTPHEADER, $headers);
+  $response = curl_exec($cURL);
+  $header_size = curl_getinfo($cURL, CURLINFO_HEADER_SIZE);
+  $response_code = curl_getinfo($cURL, CURLINFO_RESPONSE_CODE);
+  $headerSent = curl_getinfo($cURL, CURLINFO_HEADER_OUT );
+  curl_close($cURL);
+  if ($response === false) {
+    $messages = array(
+      'message' => t("Unable to receive response from curl_exec()"),
+      'message_sub' => array(),
+      'type' => 'watchdog',
+      'severity' => WATCHDOG_ERROR,
+    );
+    watchdog($messages);
+  } else {
+    $headers = substr($response, 0, $header_size);
+    $body = substr($response, $header_size);
+    if (empty($expect) || (isset($expect[$response_code]) && empty($expect[$response_code]))) {
+      return $body;
+    } else if (isset($expect[$response_code])) {
+      $expectation = $expect[$response_code];
+      $json = json_decode($body, true);
+      if (is_array($json)) {
+        if (isset($json[$expectation])) {
+          return $json;
+        } else {
+          $messages = array(
+            'message' => t("JSON key from @endpoint (@expectation) missing . @response"),
+            'message_sub' => array(
+              '@endpoint' => $endpoint, 
+              '@expectation' => $expectation,
+              '@response' => $response,
+            ),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
```

Comment:
`$failureMessage[]` is unused in this function and can be dropped from 242, 255, and 268.  The logic of this function currently only supports a single setting of the `$messages` array, via the hierarchy of `if` ... `else` statements.